chips/earlgrey/src/timer.rs

```diff
@@ -71,59 +75,121 @@ impl<'a> RvTimer<'a> {
+        // at once, which creates a race condition if the lower register
+        // wraps between the reads. So the recommended approach is to read
+        // low, read high, read low, and if the second low is lower, re-read
+        // high. -pal 8/6/20
+        let first_low: u32 = self.registers.value_low.get();
+        let mut high: u32 = self.registers.value_high.get();
+        let second_low: u32 = self.registers.value_low.get();
+        if second_low < first_low {
+            // Wraparound
+            high = self.registers.value_high.get();
+        }
+        Ticks64::from(((high as u64) << 32) | second_low as u64)
+    }
+}
 
-    fn now(&self) -> u32 {
-        self.registers.value_low.get()
+impl<'a> time::Counter<'a> for RvTimer<'a> {
+    fn set_overflow_client(&'a self, client: &'a dyn time::OverflowClient) {
+        self.overflow_client.set(client);
     }
-    fn max_tics(&self) -> u32 {
-        core::u32::MAX
+
+    fn start(&self) -> ReturnCode {
+        ReturnCode::SUCCESS
+    }
+
+    fn stop(&self) -> ReturnCode {
+        // RISCV counter can't be stopped...
+        ReturnCode::EBUSY
+    }
+
+    fn reset(&self) -> ReturnCode {
+        // RISCV counter can't be reset
+        ReturnCode::FAIL
+    }
+
+    fn is_running(&self) -> bool {
+        true
     }
 }
 
 impl<'a> time::Alarm<'a> for RvTimer<'a> {
-    fn set_client(&self, client: &'a dyn time::AlarmClient) {
-        self.client.set(client);
+    fn set_alarm_client(&self, client: &'a dyn time::AlarmClient) {
+        self.alarm_client.set(client);
     }
 
-    fn set_alarm(&self, tics: u32) {
+    fn set_alarm(&self, reference: Self::Ticks, dt: Self::Ticks) {
+        // This does not handle the 64-bit wraparound case.
+        // Because mtimer fires if the counter is >= the compare,
+        // handling wraparound requires setting compare to the
+        // maximum value, issuing a callback on the overflow client
+        // if there is one, spinning until it wraps around to 0, then
+        // setting the compare to the correct value.
         let regs = self.registers;
+        let now = self.now();
+        let mut expire = reference.wrapping_add(dt);
 
-        // Make sure that any overlow into the high bits of the timer (which we are ignoring for
-        // now) do not have an effect on the alarm.
-        regs.value_high.set(0);
+        if !now.within_range(reference, expire) {
+            expire = now;
+        }
 
-        regs.compare_low.set(tics);
-        regs.intr_enable.write(intr::timer0::SET);
+        let val = expire.into_u64();
+        let high = (val >> 32) as u32;
+        let low = (val & 0xffffffff) as u32;
+
+        // Recommended approach for setting the two compare registers
+        // (RISC-V Privileged Architectures 3.1.15) -pal 8/6/20
+        regs.compare_low.set(0xffffffff);
+        regs.compare_high.set(high);
+        regs.compare_low.set(low);
+        self.registers.intr_enable.write(intr::timer0::SET);
```

Comment:
This should also be done for the generic mtimer implementation in `arch/rv32i`, or else this will work on OpenTitan but fail on our other RISC-V chips as soon as the timer hits 2^32 ticks.

Reply:
Yes. Sorry I missed the machine_timer; we are currently only implemented on the platforms we could test.

Reply:
perhaps @bradjc can test on the Arty-e21 once the machine_timer is implemented

Reply:
@bradjc I updated rv32i `machine_timer`.